### ISSUE TITLE
ipfs add is inconvenient through the API

### ISSUE THREAD

**Issue by MichaelMure** [ipfs/go-ipfs#3962]
Sorry for the title gore, I'll update when I get a better sense of what the real problem is / what should be done.

For my project (https://github.com/MichaelMure/TotallyNotArbore) I need to add files and directories to ipfs to make them available for download by other peers. My project is in js and use go-ipfs through js-ipfs-api. I also want to use the filestore to not have to double the disk space used.

Here is what I found for now, using the API:
- even for adding data locally, files are actually read by the js code (`util.addFromFs` in `js-ipfs-api`) and streamed in the HTTP request. It take 1.1s to add a file through the CLI and around 3s to add the same file through the API. Not a huge problem but still.
- if I add a directory, I get in the response hashes and path for each child file of the directory so it needs to get parsed, so more overhead. The `quieter` option of the cli doesn't work. Still not a big problem.
- ~~as the data is streamed through the http request, the `nocopy` option of the filestore doesn't and cannot work.~~ --> apparently that works

So I thought about bypassing the API and spawning a CLI process to add the data. But apparently the `add` command unlike the others doesn't support the `--encoding=json` option so if I need more than just the final root hash (with the `quieter` option), I need to parse human-readable text.

~~The ideal solution for me would be to have a API endoint that accept a filepath and the daemon would read that itself. Obviously this could be a security problem so it would need to be restricted to a local query. That way, we get the same performance and feature as with the CLI.~~

~~I realize writing this that I can go with the spawning a process + `quieter` option for now as I only need the root hash but it's really hacky and I will have trouble quickly if I need more.~~

Thought ?

#### Version information:
go-ipfs version: 0.4.9-
Repo version: 5
System version: amd64/linux
Golang version: go1.8.1

**Comment by Kubuxu** [comment]
The CLI uses the same API you are using.
The daemon can't read the file as the file might not even be local to your machine/container (you can communicated with the daemon through network allows for use in docker and remote servers) or it might not have permission to the file.

**Comment by MichaelMure** [comment]
I did more digging, turns out I was wrong:
- the CLI does use the API only, the file content is streamed even with the --nocopy option (what for, that I don't know).
- the --nocopy option seems not supported in js-ipfs-api (opened https://github.com/ipfs/js-ipfs-api/issues/562)
- i guess the performance offset might be due to javascript, node.js or electron

**Comment by Kubuxu** [comment]
cc @kevina

**Comment by whyrusleeping** [comment]
@MichaelMure the file content is streamed even on nocopy so that we can validate the hashes, and store correct information in the filestore.

How much data are you adding such that its taking multiple seconds?

**Comment by MichaelMure** [comment]
> @MichaelMure the file content is streamed even on nocopy so that we can validate the hashes, and store correct information in the filestore.

I see ! Thanks for the enlightening.

I made some perf test again with a 100mo file from /dev/urandom. Empty repo each time. CLI timing are measured with `time`, the other with a manual timer so they are inaccurate. I also only did one measure for each.

|                         | 1st add | 2nd add |
|-------------------------|:-------:|:-------:|
| CLI                     |   7.4s   |   0.56s  |
| CLI + daemon            |   7.7s   |   1.3s   |
| Arbore dev              |    12s   |   2.3s   |
| Arbore prod             |    9s    |   1.3s   |

There is a clear slowdown when running my app in dev (maybe due to Flow ?). There is also a slowdown compared to the raw CLI add, but that can be a lot of things.

To summarize:
- the perf is not much a concern for me right now
- https://github.com/ipfs/js-ipfs-api/issues/562 prevent me from using the filestore
- the fact that the API doesn't care about the `quieter` option is annoying but it's not the end of the world.

**Comment by kevina** [comment]
This sounds like it is an issue with `js-ipfs-api` and not `go-ipfs` or am I missing something?

**Comment by whyrusleeping** [comment]
@kevina Yeah, i think its an issue with js-ipfs-api

**Comment by Kubuxu** [comment]
> the fact that the API doesn't care about the quieter option is annoying but it's not the end of the world.

It is implemented client side.

**Comment by MichaelMure** [comment]
Yes, I opened the real issues in js-ipfs-api and I'm less confused now. Closing this one. Thanks !
